### ISSUE TITLE
feat(manifolds): EuclideanManifold for unconstrained scalar parameters trained alongside gates

### ISSUE THREAD

**Issue by zazabap** [zazabap/pdft#19]
## Motivation

`pdft.manifolds` currently provides Riemannian manifold types for the constrained gate parameters that appear in QFT-family bases: `UnitaryManifold` (U(d)), `Unitary2qManifold` (U(4) on 2-qubit gates), `OrthogonalManifold` / `Orthogonal2qManifold` (O(d) variants), and `PhaseManifold` (U(1)^d, unit-complex circle). The auto-classifier `classify_manifold(t)` (loss.py is unrelated; see `manifolds.py:140-159`) routes a tensor based on shape and unitarity, and falls through to `PhaseManifold` for anything not recognised as a unitary matrix.

There is no manifold type for *unconstrained real (or complex) parameters*. This blocks a class of training problems that mix constrained gate tensors with one or a handful of free scalars — for example:

- A *learnable block-size* parameter $s \in \mathbb{R}$ jointly trained with the gate tensors of `QFTBasis(m, n)` (downstream use case in pdft-benchmarks; see "Linked downstream artefacts" below).
- Per-gate continuous mask weights that interpolate between "pinned" and "free."
- Learnable loss-temperature schedules, learnable regulariser strengths, etc.

Today these have no home: the auto-classifier sends them to `PhaseManifold`, which constrains $|s| = 1$ — wrong geometry. The downstream consumer either has to (a) train the scalar in a separate outer loop (alternating-update scheme, much slower due to JIT retracing and worse convergence due to non-coherent gradient flow), or (b) reimplement the entire batched training loop in their own package.

The smallest fix is to add an `EuclideanManifold` type whose `project` is the identity and whose `retract` is standard SGD.

## Proposed change

New manifold class in `src/pdft/manifolds.py`:

```python
@dataclass(frozen=True)
class EuclideanManifold:
    """Unconstrained Euclidean manifold (R^d or C^d), no curvature.

    For real- or complex-valued parameters that are not constrained
    to a Stiefel / unitary / phase submanifold. The Adam step is
    ordinary SGD on the entries: project is identity (full gradient
    is in the tangent space at every point), retract is x + alpha *
    xi (no manifold curvature to correct for), transport is identity
    (parallel transport in flat space is the identity map).

    Use cases: learnable scalar hyperparameters trained jointly with
    constrained gate tensors (e.g. learnable block-size, learnable
    mask weights, learnable loss-temperature schedules).
    """

    def project(self, x: Array, g: Array) -> Array:
        return g

    def retract(self, x: Array, xi: Array, alpha: float,
                *, I_batch=None) -> Array:
        return x + alpha * xi

    def transport(self, x_old: Array, x_new: Array, v: Array) -> Array:
        return v
```

It does **not** participate in `classify_manifold`'s auto-routing, because the current cascade ends at `PhaseManifold` and there is no shape signature that distinguishes "a scalar I want to leave free" from "a scalar I want on the unit circle." The user opts in explicitly when constructing the basis or by overriding the manifold for specific tensor indices — see "Design choices for discussion" below.

Tests added in `tests/optimizers/` (or wherever the existing manifold tests live):

- `EuclideanManifold.project(x, g) == g` for arbitrary shapes (real and complex).
- `EuclideanManifold.retract(x, xi, alpha) == x + alpha * xi` (no curvature correction).
- A two-step training smoke test on a toy loss $\mathcal{L}(s) = (s - 3.7)^2$ converging to $s \approx 3.7$.
- (Optional) An integration test where a basis has one tensor on `EuclideanManifold` and the rest on `UnitaryManifold`, confirming that `group_by_manifold` buckets correctly and a joint Adam step produces consistent updates.

## Design choices for discussion

1. **Auto-routing in `classify_manifold`.**

   - *Option (a)*: do not auto-route to `EuclideanManifold`. Force the user to opt in explicitly. Keeps backward compatibility (no existing code accidentally lands in `EuclideanManifold` instead of `PhaseManifold`).
   - *Option (b)*: auto-route real-valued, non-unitary tensors to `EuclideanManifold`. This would change the current behaviour for any real-valued non-unitary tensor that was implicitly going to `PhaseManifold` and being held to $|x| = 1$ — a behavioural change.

   **Recommendation:** Option (a). Backward-compatible. The user opts in by either (i) explicitly assigning the manifold for specific indices via a basis-class override hook (if/when one exists), or (ii) wrapping the scalar in a marker class that the classifier recognises.

2. **How does a basis class declare "this tensor goes on `EuclideanManifold`"?**

   The current `classify_manifold` is purely shape-based — it has no notion of "the user said keep this one free." For sibling PR `_extra_loss` (PR #18), the loss carries its own subclass identity, which is enough to add the reg term. But here, the *manifold assignment* is intrinsic to the basis: it lives in `group_by_manifold(basis.tensors)`.

   - *Option (a)*: A marker wrapper class around the scalar tensor that `classify_manifold` recognises and routes to `EuclideanManifold`. Lightweight; backward-compatible.
   - *Option (b)*: A new method on basis classes, `manifold_for(index) -> AbstractRiemannianManifold | None`, that overrides the default classifier for specific indices.
   - *Option (c)*: A new constructor argument on `QFTBasis(m, n, ..., free_indices: Set[int])` that pins specific positions to `EuclideanManifold`.

   **Recommendation:** *Discuss.* Option (b) is the most general; option (c) is the most explicit; option (a) is the most minimal. The right call depends on how many downstream use cases will need this and how heavyweight the per-index override needs to be. For the single learnable-$s$ use case, option (a) suffices.

3. **Name: `EuclideanManifold` vs alternatives.**

   - `EuclideanManifold` — neutral, geometric. Suggests "the Euclidean metric space" (matches the math).
   - `RealManifold` — wrong, because the Euclidean manifold is just as valid for complex parameters.
   - `FlatManifold` — accurate (zero curvature) but vague.
   - `UnconstrainedManifold` — accurate but operationally negative ("not constrained").
   - `TrivialManifold` — concise but obscures the role.

   **Recommendation:** `EuclideanManifold`. Matches the Riemannian-geometry naming convention used elsewhere in the module (Unitary*, Orthogonal*, Phase*).

4. **Type signature for `project`.**

   The protocol's `project(points, grads) -> Array` returns the "Riemannian gradient" given a Euclidean gradient. For Euclidean manifolds, this is the identity map. But should `project` take a complex gradient and return a real-valued one when the parameter is intended to be real? *Recommendation:* keep it generic — accept and return whatever dtype `g` arrives in. Real-valuedness is a property of the *parameter type*, enforced by the dtype of `x`, not by the manifold.

## Scope

- Adds ~25 LoC to `src/pdft/manifolds.py` (the new class).
- Adds ~3-4 tests to `tests/optimizers/`.
- Does NOT modify `classify_manifold` (per design choice 1, recommendation (a)).
- Does NOT add a routing hook on basis classes (per design choice 2; revisit if multiple downstream use cases need it).

## Out of scope for this issue

- Constrained-box manifolds (e.g., $[a, b]$ enforced by reflection). The standard reparametrisation trick $x = a + (b - a) \cdot \sigma(\tilde{x})$ with $\tilde{x} \in \mathbb{R}$ handles this downstream; no upstream support needed.
- Per-basis-index manifold-override API. Discuss in design choice 2; defer until needed.
- The downstream learnable-$s$ regulariser itself — lives in `pdft-benchmarks`.

## Acceptance criteria

- [ ] Existing `tests/manifolds/` / `tests/optimizers/` tests still pass.
- [ ] New tests covering `project` / `retract` / `transport` for `EuclideanManifold`.
- [ ] An integration test where a basis mixes `UnitaryManifold` and `EuclideanManifold` tensors and a single Adam step updates both correctly.
- [ ] Full pdft test suite (`pytest tests/`) passes.

## Linked downstream artefact (for context, not in scope)

- pdft-benchmarks `results/qft_identity_init/writeup_reg_sweep.pdf` — "Proposed follow-up: learnable block size" section, which motivates this issue.
- The downstream consumer wants to train a scalar $s$ (block-size parameter) jointly with the 72 gate tensors of `QFTBasis(8, 8)` via a single Adam step. With `EuclideanManifold` available, this fits inside the existing `pdft.train_basis_batched` API with no further changes.

**Comment by zazabap** [comment]
**Status update (2026-05-13):** the immediate downstream consumer that motivated this issue — a "learnable block size" regulariser in pdft-benchmarks — has been deprioritised in favour of a simpler L1-to-identity regulariser that requires no upstream changes.

The L1 variant uses standard `MSELoss._extra_loss` (PR #18) and trains entirely on the existing U(2) gate manifold; sparsity emerges from the L1 dynamics without any new scalar parameter. This addresses the same underlying research question (whether a matched structural prior is *necessary* to find the `blocked_8` basin, or whether plain sparsity suffices) more directly and with less machinery.

`EuclideanManifold` itself remains generally useful — any future "mixed constrained gate parameters + free scalars" use case is still blocked without it. Recommend leaving this issue open as a backlog item; revisit when a concrete second use case appears.

(Context: `pdft-benchmarks` `results/qft_identity_init/writeup_reg_sweep.pdf` — "Proposed follow-up: L1-to-identity" section, replacing the earlier learnable-$s$ proposal that referenced this issue.)
